File: transpiler/javatests/com/google/j2cl/integration/java/supermethodcall/Main.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package supermethodcall;

import static com.google.j2cl.integration.testing.Asserts.assertEquals;

/**
 * Test non-constructor super method calls.
 */
public class Main {
  public static void main(String[] args) {
    testSuperMethodCalls();
    testSuperMethodCalls_returnSpecialization();
    testSuperMethodCalls_parameterSpecialization();
    testSuperMethodCalls_defaultMethods();
  }

  private static void testSuperMethodCalls() {
    class GrandParent {
      public String getName() {
        return "GrandParent";
      }
    }

    class Parent extends GrandParent {
      public String getName() {
        return "Parent";
      }

      public String getParentName() {
        return super.getName();
      }
    }

    class Child extends Parent {
      @Override
      public String getName() {
        return "Child";
      }

      @Override
      public String getParentName() {
        return super.getName();
      }

      public String getGrandParentName() {
        return super.getParentName();
      }

      class Inner {
        public String getOuterParentName() {
          return Child.super.getName();
        }
      }
    }
    assertEquals("Parent", new Child().getParentName());
    assertEquals("GrandParent", new Child().getGrandParentName());
    assertEquals("Parent", new Child().new Inner().getOuterParentName());
  }

  private static void testSuperMethodCalls_returnSpecialization() {
    class GrandParent {
      public CharSequence getName() {
        return "GrandParent";
      }
    }

    class Parent extends GrandParent {
      @Override
      public String getName() {
        return "Parent";
      }

      public String getParentName() {
        return (String) super.getName();
      }
    }

    class Child extends Parent {
      public String getName() {
        return "Child";
      }

      public String getParentName() {
        return super.getName();
      }

      public String getGrandParentName() {
        return super.getParentName();
      }
    }

    assertEquals("Parent", new Child().getParentName());
    assertEquals("GrandParent", new Child().getGrandParentName());
  }

  private static void testSuperMethodCalls_parameterSpecialization() {
    class GrandParent<T> {
      public T getName(T object) {
        return (T) "GrandParent";
      }
    }

    class Parent<T extends CharSequence> extends GrandParent<T> {
      @Override
      public T getName(T object) {
        return (T) "Parent";
      }

      public T getParentName(T object) {
        // Dispatches to an override.
        return super.getName(object);
      }
    }

    class Child extends Parent<String> {
      // Parameter Specialization
      public String getName(String object) {
        return "Child";
      }

      public String getParentName(String object) {
        return super.getName(object);
      }

      public String getGrandParentName(String object) {
        return super.getParentName(object);
      }
    }

    assertEquals("Parent", new Child().getParentName(null));
    assertEquals("GrandParent", new Child().getGrandParentName(null));
  }

  interface GrandParentInterface {
    default String getName() {
      return "GrandParent";
    }

    default String getDiamondName() {
      return "Diamond";
    }
  }

  interface Parent1Interface extends GrandParentInterface {
    default String getName() {
      return "Parent1";
    }

    default String getParentName() {
      return GrandParentInterface.super.getName();
    }
  }

  interface Parent2Interface extends GrandParentInterface {
    default String getName() {
      return "Parent2";
    }

    default String getParentName() {
      return GrandParentInterface.super.getName();
    }
  }

  interface ChildInterface extends Parent1Interface, Parent2Interface {
    default String getName() {
      return "Child";
    }

    default String getParentName() {
      return Parent1Interface.super.getName() + "+" + Parent2Interface.super.getName();
    }

    default String getGrandParentName() {
      return Parent1Interface.super.getParentName() + "+" + Parent2Interface.super.getParentName();
    }
  }

  interface GrandChildInterface extends ChildInterface {
    default String getDiamondName() {
      return ChildInterface.super.getDiamondName();
    }
  }

  private static void testSuperMethodCalls_defaultMethods() {
    assertEquals("GrandParent", new GrandParentInterface() {}.getName());
    assertEquals("GrandParent", new Parent1Interface() {}.getParentName());
    assertEquals("Parent1", new Parent1Interface() {}.getName());
    assertEquals("GrandParent", new Parent2Interface() {}.getParentName());
    assertEquals("Parent2", new Parent2Interface() {}.getName());
    assertEquals("GrandParent+GrandParent", new ChildInterface() {}.getGrandParentName());
    assertEquals("Parent1+Parent2", new ChildInterface() {}.getParentName());
    assertEquals("Child", new ChildInterface() {}.getName());
    assertEquals("Diamond", new GrandChildInterface() {}.getDiamondName());
  }
}

File: transpiler/javatests/com/google/j2cl/integration/kotlin/supermethodcall/Main.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package supermethodcall

import com.google.j2cl.integration.testing.Asserts.assertEquals

/** Test non-constructor super method calls. */
fun main(vararg unused: String) {
  testSuperMethodCalls()
  testSuperMethodCalls_returnSpecialization()
  testSuperMethodCalls_parameterSpecialization()
  testSuperMethodCalls_defaultMethods()
}

private fun testSuperMethodCalls() {
  open class GrandParent {
    open fun getName(): String? {
      return "GrandParent"
    }
  }

  open class Parent : GrandParent() {
    override fun getName(): String? {
      return "Parent"
    }

    open fun getParentName(): String? {
      return super.getName()
    }
  }

  class Child : Parent() {
    override fun getName(): String? {
      return "Child"
    }

    override fun getParentName(): String? {
      return super.getName()
    }

    fun getGrandParentName(): String? {
      return super.getParentName()
    }

    inner class Inner {
      fun getOuterParentName(): String? {
        return super@Child.getName()
      }
    }
  }
  assertEquals("Parent", Child().getParentName())
  assertEquals("GrandParent", Child().getGrandParentName())
  assertEquals("Parent", Child().Inner().getOuterParentName())
}

private fun testSuperMethodCalls_returnSpecialization() {
  open class GrandParent {
    open fun getName(): CharSequence? {
      return "GrandParent"
    }
  }

  open class Parent : GrandParent() {
    override fun getName(): String? {
      return "Parent"
    }

    open fun getParentName(): String? {
      return super.getName() as String
    }
  }

  class Child : Parent() {
    override fun getName(): String? {
      return "Child"
    }

    override fun getParentName(): String? {
      return super.getName()
    }

    fun getGrandParentName(): String? {
      return super.getParentName()
    }
  }

  assertEquals("Parent", Child().getParentName())
  assertEquals("GrandParent", Child().getGrandParentName())
}

private fun testSuperMethodCalls_parameterSpecialization() {
  open class GrandParent<T> {
    open fun getName(o: T): T {
      return "GrandParent" as T
    }
  }

  open class Parent<T : CharSequence?> : GrandParent<T>() {
    override fun getName(o: T): T {
      return "Parent" as T
    }

    open fun getParentName(o: T): T {
      // Dispatches to an override.
      return super.getName(o)
    }
  }

  class Child : Parent<String?>() {
    // Parameter Specialization
    override fun getName(o: String?): String? {
      return "Child"
    }

    override fun getParentName(o: String?): String? {
      return super.getName(o)
    }

    fun getGrandParentName(o: String?): String? {
      return super.getParentName(o)
    }
  }

  assertEquals("Parent", Child().getParentName(null))
  assertEquals("GrandParent", Child().getGrandParentName(null))
}

interface TopInterface {
  fun getName(): String {
    return "Top"
  }

  fun getDiamondName(): String {
    return "Diamond"
  }
}

interface LeftInterface : TopInterface {
  override fun getName(): String {
    return "Left"
  }

  fun getSuperName(): String {
    return super.getName()
  }
}

interface RightInterface : TopInterface {
  override fun getName(): String {
    return "Right"
  }

  fun getSuperName(): String {
    return super.getName()
  }
}

interface BottomInterface : LeftInterface, RightInterface {
  override fun getName(): String {
    return "Bottom"
  }

  override fun getSuperName(): String {
    return super<LeftInterface>.getName() + "+" + super<RightInterface>.getName()
  }

  fun getSuperSuperName(): String {
    return super<LeftInterface>.getSuperName() + "+" + super<RightInterface>.getSuperName()
  }
}

interface GrandBottomInterface : BottomInterface {
  override fun getDiamondName(): String {
    return super.getDiamondName()
  }
}

private fun testSuperMethodCalls_defaultMethods() {
  assertEquals("Top", object : TopInterface {}.getName())
  assertEquals("Top", object : LeftInterface {}.getSuperName())
  assertEquals("Left", object : LeftInterface {}.getName())
  assertEquals("Top", object : RightInterface {}.getSuperName())
  assertEquals("Right", object : RightInterface {}.getName())
  assertEquals("Top+Top", object : BottomInterface {}.getSuperSuperName())
  assertEquals("Left+Right", object : BottomInterface {}.getSuperName())
  assertEquals("Bottom", object : BottomInterface {}.getName())
  assertEquals("Diamond", object : GrandBottomInterface {}.getDiamondName())
}
